agent/drive_handler.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/firecracker-microvm/firecracker-containerd/internal"
+)
+
+const (
+	blockPath       = "/sys/block"
+	drivePath       = "/dev"
+	blockMajorMinor = "dev"
+)
+
+type drive struct {
+	Name       string
+	DriveID    string
+	MajorMinor string
+	DrivePath  string
+}
+
+type driveHandler struct {
+	// drives is a mapping to all the stub drives
+	drives map[string]drive
+	// BlockPath is the path which lists all block devices. In Linux this is
+	// typically /sys/block
+	BlockPath string
+	// DrivePath is the directory where the list of drives will be found. In
+	// Linux this is typically /dev
+	DrivePath string
+}
+
+func newDriveHandler(blockPath, drivePath string) (*driveHandler, error) {
+	d := &driveHandler{
+		drives:    map[string]drive{},
+		BlockPath: blockPath,
+		DrivePath: drivePath,
+	}
+
+	err := d.discoverDrives()
+	if err != nil {
+		return nil, err
+	}
+
+	return d, nil
+}
+
+func (dh driveHandler) GetDrive(id string) (drive, bool) {
+	v, ok := dh.drives[id]
+	return v, ok
+}
+
+// discoverDrives will iterate the block path in the sys directory to retrieve all
+// stub block devices.
+func (dh *driveHandler) discoverDrives() error {
+	names, err := getListOfBlockDeviceNames(dh.BlockPath)
+	if err != nil {
+		return err
+	}
+
+	drives := map[string]drive{}
+	for _, name := range names {
+		d, err := dh.buildDrive(name)
+		if err != nil {
+			return err
+		}
+
+		if !isStubDrive(d) {
+			continue
+		}
+
+		f, err := os.Open(d.Path())
+		if err != nil {
+			return err
+		}
+
+		d.DriveID, err = internal.ParseStubContent(f)
+		f.Close()
```

Comment:
Don't ignore the return value of `f.Close()` There aren't many ways it can fail on a read-only fd, but I'd rather be sure than potentially miss something.